storage/storage.go

```diff
@@ -1057,6 +1058,37 @@ func (o *ObjectHandle) OverrideUnlockedRetention(override bool) *ObjectHandle {
 	return &o2
 }
 
+// SoftDeleted returns an object handle that can be used to get an object that
+// has been soft deleted. To get a soft deleted object, the generation must be
+// set on the object using ObjectHandle.Generation.
+// Note that an error will be returned if a live object is queried using this.
+func (o *ObjectHandle) SoftDeleted() *ObjectHandle {
+	o2 := *o
+	o2.softDeleted = true
+	return &o2
+}
+
+// Restore will restore a soft-deleted object to full object status.
+// Note that you must specify a generation to use this method.
+// copySourceACL indicates whether the restored object should copy the
+// access controls of the source object.
+func (o *ObjectHandle) Restore(ctx context.Context, copySourceACL bool) (*ObjectAttrs, error) {
+	if err := o.validate(); err != nil {
+		return nil, err
+	}
+	// Restore is idempotent if GenerationMatch or Generation have been passed in.
+	// The default generation is negative to get the latest version of the object.
+	isIdempotent := (o.conds != nil && o.conds.GenerationMatch != 0) || o.gen >= 0
```

Comment:
It says in the function docs that a generation is required. If so can we skip this check?

Reply:
I think that's fine, but want to check in with @JesseLovelace - does this check get skipped in other languages?

Reply:
Okay, let's check this before merging. It would be better if this were not required.

Reply:
Yeah I don't think we have that in the other languages